### ISSUE TITLE
Threads and Memory options

### ISSUE THREAD

**Issue by sejmodha** [PoonLab/OpenRDP#32]
Hi There

I am using OpenRDP pipeline on Linux for a reasonably large dataset of around 1200 sequences and noticed that all server resources (all threads, memory+ swap) are being utilised by OpenRDP after 3Seq analysis is completed. I could not find any option that allows me to limit the resources being used by the pipeline eg. the number of threads or maximum memory..

It would be great if control options for resources were provided at the top level of the pipeline such that it limits the resources for all tools.

Thanks
Sej

**Comment by ArtPoon** [comment]
Could you please copy your command line invocation to this issue?  Were you only calling the 3Seq method?

**Comment by sejmodha** [comment]
Here is the abstract command, I am running the entire pipeline with `-all`
`python3 -m openrdp {input} {output} -all`

**Comment by ArtPoon** [comment]
You mentioned that resource usage went up *after* the 3Seq analysis completed - do you know if that was associated with running the next method in the suite, *i.e.*, GENECONV?  Please bear with me - my former student @kwade4 did the bulk of development for this program so it's slower for me to attend to your issue.

**Comment by sejmodha** [comment]
No worries at all....any help is very much appreciated. 

So far I have got `3s.log`,  `3s.longRec`, `3s.pvalHist`,  `3s.rec`  files that are generated by 3Seq. 

However, the standard out log captured in a file suggest that GENECONV analysis is finished (see above comment). 
Does that log also suggest that it may be the Bootscan/Rescan analysis that may have caused the memory and threads hogging? 
Moreover, I found it a bit strange that I am unable to see what step of the process is currently being executed as I can only see the original `openrdp` command being executed without details of what stage the pipeline is at...

Hope that helps. 

Thanks once again.

**Comment by darrenpmartin** [comment]
Hi All - it is very plausible that bootscan is the resource hog - it scales
terribly with sequence number (lots of n x n matrices).  To handle >1000
sequences bootscan would need to have some sort of divide and
conqueror based downsampling scheme.  with 1K sequences amd 100K bootstrap
reps at each step bootscan would generate 800Megs worth of distance
matrices AT EACH STEP.  If it is set up to run with 200 steps it will maybe
be trying to store 200 X 800 megs = 150 Gigs worth of distance matrix data.

regards
darren

**Comment by ArtPoon** [comment]
Thanks for the insight @darrenpmartin!  Should be straight-forward for us to implement a maximum limit and downsample to deal with large inputs.  I'll also check the Python script to see if there are any easy optimizations.

**Comment by ArtPoon** [comment]
I haven't forgotten about this issue, but I'm scrambling to deal with a bunch of other repositories and the main package author has moved on.

**Comment by ArtPoon** [comment]
I think first step is to run some basic tests on memory consumption for `bootscan.py`

**Comment by ArtPoon** [comment]
Let's start with measuring time and memory complexity for small to large data sets, and then go into optimizing `bootscan.py` @dpannguyen

**Comment by dpannguyen** [comment]
Real time and peak memory usage were measured for data set of 1, 10, and 50 sequences as below:

dataset (seq) | method | real time (s) | peak memory usage (MB)
:---: | :--- | ---: | ---:
1 | `threeseq` | 1.33 | 75.8
1 | `geneconv` | 1.43 | 75.8
1 | `maxchi` | 1.43 | 75.9
1 | `chimaera` | 1.36 | 75.8
1 | `bootscan` | **_2.72_** | **_79.9_**
1 | `siscan` | 1.35 | 75.7
1 | `rdp` | 1.41 | 75.9
| | | 
10 | `threeseq` | 2.38 | 88.5
10 | `geneconv` | 2.48 | 75.9
10 | `maxchi` | 12.11 | 76.9
10 | `chimaera` | 7.76 | 76.7
10 | `bootscan` | **_38.12_** | **_100.1_**
10 | `siscan` | 5.57 | 76.5
10 | `rdp` | 5.62 | 76.7
| | | 
50 | `threeseq` | 11.19 | 90.5
50 | `geneconv` | 26.48 | 76.3
50 | `maxchi` | 1635.52 | 78.7
50 | `chimaera` | 950.35 | 80.6
50 | `bootscan` | **_3387.03_** | **_515_**
50 | `siscan` | 551.46 | 79.6
50 | `rdp` | 502.85 | 84.3


`bootscan` is definitely the method that takes up the most resource and it scales up significantly as the dataset grows (taking nearly an hour to process 50 sequences in this case). As @darrenpmartin mentioned, we will look into optimizing the code with a divide and conquer method.

**Comment by darrenpmartin** [comment]
Hi An - How did you manage to run a dataset of 1 sequence? the minimum
should be 3. I am also surprised that the runtime and memory usage of
siscan is similar to that of rdp, chimaera and maxchi. These latter
methods  "should" be at least as fast as geneconv.
regards
Darren

**Comment by dpannguyen** [comment]
@darrenpmartin Hi Darren, you were absolutely right. That was my bad, I wrote a short script to run and grab the runtime and memory usage of each of the methods. Since the specs still showed up even after the process got terminated by the error, it accidentally grabbed that for the dataset of 1 as well. I have updated the table above with the new values for 3 sequences.

Regarding the runtime for `rdp`, `chimaera`, and `maxchi`, we can probably have a look at them after `bootscan` optimization. Thank you!

**Comment by darrenpmartin** [comment]
Hi An - Are you shelling to geneconv and 3seq?  If so then the difference
may just be that between c++ and python.

**Comment by ArtPoon** [comment]
@darrenpmartin - I'm pretty sure that's the case, that the methods that were ported into Python need optimizing.  I haven't had a chance to go through the code and @dpannguyen has very kindly volunteered to help out with bootscan

**Comment by ArtPoon** [comment]
and yes, we're using `subprocess` calls out to 3seq and geneconv binaries.

**Comment by ArtPoon** [comment]
> OK so geneconv is the originally stanley sawyer version? That means it does not do triplet scans, right? Or did Stanley add that in the last few years If it is just pairwise scanning that would explain part of the massive performance difference too.
> […](#)

As far as I recall, @kwade4 obtained the original geneconv binaries.  Is there a customized version bundled with RDP?  As for added features, my understanding is that Dr. Sawyer unfortunately passed away in 2017...

**Comment by darrenpmartin** [comment]
Damn - I didn't know that Stanly had died.  I last spoke to him ~10 years
ago. The version bundled with RDP is just some old legacy thing - RDP uses
it to draw plots while it is doing other stuff (its a clumsy sort of
parallelization) but doesnt use it for the geneconv tests that RDP
performs. For that RDP uses a rewritten version of the geneconv code that
runs quickly enough that it can be used to scan triplets.Its not a biggie
though - I'm sure the pairwise scanning version of the original will work
fine (and a lot faster than a triplet scanning version).

**Comment by GopiGugan** [comment]
I ran a profiler (using a fasta file with 10 sequences) to identify where we could make some improvements:
1) Run time: 
https://github.com/PoonLab/OpenRDP/blob/3a59bc1ae09c7fed4af4dd077c02d5abe9e097ad/openrdp/scripts/bootscan.py#L149 
This operation takes roughly 48% of the time
2) Memory:
https://github.com/PoonLab/OpenRDP/blob/3a59bc1ae09c7fed4af4dd077c02d5abe9e097ad/openrdp/scripts/bootscan.py#L128
Total memory usage: 594.3 MB
Storing the distance matrix requires 498.4 MB
As mentioned earlier, this step would require a lot more memory when using a fasta file with more sequences

**Comment by darrenpmartin** [comment]
I assume you mean storing all the distance matrices?  If so then yes
storing all of the distance matrices (num windows scanned x num bootstraps)
will take up a lot of space.  That is why the calced distances should all
get dropped to the disk rather than holding them in ram.  If you store the
numbers properly (i.e. not as distance matrices but rather as lists of
distances that, for any sequence pair, can be loaded as a single block from
the disk, the cost in read-time is minimal.  You just need to sort the
distances properly before you drop them to the disk.  The temporary file on
disk will, of course, get very large for even a few hundred sequences.
Obviously you need to calc distances for all windows and bootstrap
replicates for every pair of sequences one at a time which means you dont
do the bootstrap replicates like in phylip - you just keep a record of
which alignment columns occur in each replicate and you do the distance
calcs from the primary alignment using only those alignment colums (i.e.
you dont make 100 new alignments for 100 bootstrap replicates)..

**Comment by ArtPoon** [comment]
I noticed that this data structure seems to not only be sparse (containing many zeros) but also a very small number of numerical values.  It might be more efficient to store integer indices for these values, and a map of indices to the original numbers.

**Comment by darrenpmartin** [comment]
Great idea! Even doing it naively, you could do this for datasets of up to
~250 sequences right? (the square root of 64k unique pairwise distances).
You dont need 64K unique numbers though to encode the matrix - not if you
encode it as a tree that is (i.e. where, for 250 sequences, there are 248
"nodes" and you could store the sequences as sets of non-unique
"treedistances" where all the sequences in one clade have the same distance
to all the sequences in another non-overlapping clade)  You would not even
need to decode the distances since all you care about is whether sequence A
is more closely related to sequence B than it is to sequence C - the actual
distances are irrelevant as long as the encoded distances between sequences
have the same ranks as the real distances.

 So..... the lowest "treedistance" between two sequences would get an
unsigned short int value of 0, the next lowest a value of 1 etc etc. and
the whole set of relevant relationships for your 250 sequences would
compress down to just 250+248 numbers which you could store in a
dynamically manufactured 9 bit variable (stored in code within a declared
unsigned short integer array memory block big enough to fit the 9bit
numbers).  You just increase or decrease the bit-size of this manufactured
variable to cater for the dataset size such that for a 16k sequence dataset
you would reach the point where the manufactured variable would be
equivalent to a an unsigned short integer  - To make things practical you
would have individual decoders for each manufactured variable between, say,
6 bit and 32 bit (where the 32 bit decoder would be for tree distances for
datasets of millions of sequences) and you'd just store the data in the
same unsigned integer array memory block as always - on-the-fly choosing
whatever decoder you needed to convert the numbers in that array to the
distance ranks.

If you coupled a scheme like this with the dynamic storage and access of
the bootstrapped distance data from disk scheme that I currently use in
rdp, the memory footprint would be negligible even for datasets of 10s of
millions of sequences with bootscan involving 1000s of bootstrap
replicates  (since you would never even ever need to have whole distance
matrices in memory - the only compressed distance rank scores you would
need would be those of whatever triplets of sequence were pertinent for
whatever bootscan plot you were currently working on.

**Comment by GopiGugan** [comment]
Results for the current OpenRDP version (`master` branch) when running the Bootscan method:

Number of Sequences | Execution Time (s) | Memory Consumption (MB)
:---: | ---: | ---:
10 | 13.33 | 124.3
50 | 430.21 | 288.7
100 | 3243.42 | 603.8

Some of the changes that I have made:

1. Allow the user to specify the number of processes to use for the Bootstrap method in the config file. 
2. Instead of storing the entire distance matrix, store the condensed distance matrix:

    For example:
    ```python
    array([1, 2, 3, 4, 5, 6])
    ```
    Full distance matrix:

    ```python
    array([[0, 1, 2, 3],
           [1, 0, 4, 5],
           [2, 4, 0, 6],
           [3, 5, 6, 0]])
    ``` 

    Number of Sequences | Execution Time (s) | Memory Consumption (MB)
    :---: | ---: | ---:
    10 | 13.45 | 117.9
    50 | 530.74 | 205.1
    100 | 3637.39 | 450.2

3. Write the condensed distance matrices to a file using `h5py`

    Number of Sequences | Execution Time (s) | Memory Consumption (MB)
    :---: | ---: | ---:
    10 | 15.67 | 124.9
    50 | 585.53 | 139.7
    100 | 4327.97 | 145.4

4. Run the `execute()` function from `bootstrap.py` in parallel and change the data type for the values in the condensed matrix from `float64` to `float16` :
https://github.com/PoonLab/OpenRDP/blob/badc20180e5b3f621ccbd46a58ec0fc13bce1a7d/openrdp/scripts/bootscan.py#L158

    Number of Sequences | Execution Time (s)
    :---: | ---: 
    10 | 4.61
    50 | 100.84 
    100 | 1647.74

**Comment by ArtPoon** [comment]
Thanks @GopiGugan!  It looks like changing to `float16` will limit entries in the distance matrix to a maximum of 3 significant digits, e.g., 0.00123.  We should check whether this causes problems, especially for longer sequence inputs.

**Comment by ArtPoon** [comment]
Actually I'm just going to make that a new issue and deal with it later

**Comment by ArtPoon** [comment]
@GopiGugan's re-implementation of bootscan is now in main branch, along with my refactoring of the user interface
Need to update the documentation

**Comment by sejmodha** [comment]
Thanks for looking into this and addressing the optimisation. Your help is much appreciated.

**Comment by ArtPoon** [comment]
New implementation of bootscan seems to be overestimating number of recombination breakpoints in our unit tests, see #48

**Comment by darrenpmartin** [comment]
Are you sure you are using an appropriate window size? Also, if you are
using a bootstrap cutoff and forgoing any additional multiple testing
correctable tests this is a completely expected outcome.  It is absolutely
necessary to use either a chi square or binomial distribution test (or some
equivalent) that yields some or other p-value that is correctable using
Bonferroni, DS, step down or some other correction.
